src/storage/txn/actions/check_txn_status.rs

```diff
@@ -49,20 +185,23 @@ pub fn check_txn_status_lock_exists(
     }
 
     let is_pessimistic_txn = !lock.for_update_ts.is_zero();
-    if lock.ts.physical() + lock.ttl < current_ts.physical() {
-        // If the lock is expired, clean it up.
-        // If the resolving and primary key lock are both pessimistic locks, just unlock
-        // the primary pessimistic lock and do not write rollback records.
-        return if resolving_pessimistic_lock && lock.lock_type == LockType::Pessimistic {
-            let released = txn.unlock_key(primary_key, is_pessimistic_txn, TimeStamp::zero());
-            MVCC_CHECK_TXN_STATUS_COUNTER_VEC.pessimistic_rollback.inc();
-            Ok((TxnStatus::PessimisticRollBack, released))
-        } else {
-            let released =
-                rollback_lock(txn, reader, primary_key, &lock, is_pessimistic_txn, true)?;
-            MVCC_CHECK_TXN_STATUS_COUNTER_VEC.rollback.inc();
-            Ok((TxnStatus::TtlExpire, released))
-        };
+    if lock.is_pessimistic_lock() {
+        let (status, released) = check_txn_status_from_pessimistic_primary_lock(
+            txn,
+            reader,
+            primary_key.clone(),
+            &lock,
+            current_ts,
+            resolving_pessimistic_lock,
+        )?;
+        // Return if the primary lock is stale or the transaction status is decided.
+        if (status.is_decided() || status == TxnStatus::PessimisticRollBack) && released.is_some() {
+            return Ok((status, released));
```

Comment:
Can we assert `released` is empty if we are not returning here? We should be careful not to write the same thing to `MvccTxn` twice or put the same thing to `ReleasedLocks` twice.